### ISSUE TITLE
Colorize Command Clears Author Data and Paragraph-Level History from Page

### ISSUE THREAD

**Issue by dwillis77** [stevencohn/OneMore#1706]
# Problem to Solve

When we colorize text with the `colorize` command, it seems the `import` action that this triggers will clear the author data from the page. For example, if another author had authored some notes on the page and their initials were previously shown next to it to indicate this (assuming the option to "hide authors" is not enabled), that will be cleared and the current user will be marked as the author of the entire page.

Perhaps more importantly, this also messes with the page history functionality - with author data intact, each paragraph also has its own "last modified" data and thus when we look at page versions we can see changes at the paragraph level (e.g. if we changed one paragraph on a page today and we show page history, it will show only that paragraph as being modified). When this information is lost on page import, the page history for today's modification will show the entire page as being modified even if only one line or paragraph was changed.

# To Reproduce

1. Select some text
2. Go to OneMore menu -> Colorize menu and click language of choice
3. See that after the operation completes, author tags are gone and page history shows the entire page in green to indicate the entire page has been modified, even though only one small part was actually changed.

# Expected Behavior

Author data and paragraph modification data should either remain (at minimum) or, even better would be to provide an option in settings for the user to decide whether author data should be retained on page import. I think this same behavior would occur with any modification of a page (since all page modifications would involve the same "import" action, presumably) but I have not tested this to confirm.

# Environment (if applicable)

 - OneNote Version: Office 365 version 2402 (build 17328.20648) semi-annual channel, 64-bit
 - OneMore Version: 6.6.1

**Comment by stevencohn** [comment]
See https://learn.microsoft.com/en-us/answers/questions/1119987/onenote-interop-api-updatepagecontent-changes-last

**Comment by dwillis77** [comment]
Ah interesting, so it is a known issue. I appreciate the quick response!

As a workaround, I have an idea that came to mind though I'm not sure if it would actually work - if the author and date modified attributes from a page were manually stored and re-inserted back into the XML that is ultimately passed back to the `UpdatePageContent()` call - would that work to preserve it?

**Comment by stevencohn** [comment]
@dwillis77, no that's in fact the problem. The `UpdatePageContent()` call does a lot of "normalizing" in the XML and this is one area where it's obviously a little over zealous.

**Comment by dwillis77** [comment]
Ok got it, that makes sense. Thanks for explaining that! Hopefully MS gets around to fixing this one day...

**Comment by stevencohn** [comment]
Closing since this has been a known bug with OneNote for a while

**Comment by dwillis77** [comment]
@jasonjac2 FWIW, I did submit a feedback about this through OneNote's built-in tool and also upvoted your comment on the linked thread. I'm not holding my breath, although at least one other issue that I reported through that "feedback" channel in the past did actually get fixed eventually (I guess enough people complained about it to make MS take notice) so who knows maybe this will be one day as well.

@stevencohn  One side note - I did see that this doesn't seem to _always_ happen. It does the vast majority of the time but sometimes it does not. I was trying to figure out what the difference was on the pages where the issue didn't happen, and your post on the linked thread gave me a clue - you mentioned "_Given a page with one outline..._" - I think that is the key. If the page has more than one outline object in it, then the history and author info is maintained properly (at least, from what I've seen on the couple pages I looked at where the history did not get wiped after an `import` - I haven't done thorough testing to confirm this though). Does this line up with what you've seen?

If so, I suppose one manual workaround for a user would be to just add a new empty outline at the bottom of every page although that would get to be a hassle to do across many pages. And any sort of workaround that involved the add-in adding an unnecessary "dummy" outline at the bottom of the page would be a very messy solution obviously...so maybe that doesn't help much. Just thought it was worth noting though.

**Comment by stevencohn** [comment]
@dwillis77, your theory on Outlines is close. The smallest object on a page that can be updated by the OneNote Interop API is the Outline. You cannot update just a single paragraph or other object contained inside an Outline. You can update individual Outlines on a page without affecting other Outlines. 

OneMore takes this a step further. When it reads a page, it creates a temporary hash on each Outline. When it's about to save the page, it uses the hash to detect with Outlines have been updated and removes those that haven't from the XML to ensure they don't get affected - and it's a huge performance improvement when there are many Outlines. (Same for other top level objects, like InkDrawing.)

So I suspect you're seeing the untouched Outlines not being affected due to OneMore streamlining the Save operation.

**Comment by dwillis77** [comment]
Interesting @stevencohn thanks for the insight! That's pretty cool that OM has its own method to save more efficiently.

It is strange though, indeed - because in fact when I see this it isn't that some outlines are untouched (aside from the modified one) - it's that the modified one itself is untouched (aside from the change made, whatever it is) - but the rest of the paragraph-level author and history info remains intact. And this happens whether the change is made via OneMore or another plugin (that uses the same method/only method of updating page content programmatically via the interop API - i.e. `UpdatePageContent()`).

I know this is a long-standing bug so I'm not sure how much time it's worth investing in it, though it would be awesome if we could find a workaround. There's got to be some difference between the outlines that get their history and author info wiped on update, and those that don't. If it helps I'd be happy to send over page XML of a page that does not have the problem, for comparison via a page that does (to be clear - it's a tiny percentage - maybe one out of every 30-50 pages I update via plugin will retain the author and history info). Otherwise I may try to dig into it further myself, when I get a chance.

Thanks again for all your help.

**Comment by dwillis77** [comment]
Just want to post one more follow-up here in case anyone else comes across this with the same sort of questions I had. After further testing, the above scenario I mentioned where we modify an outline and the page history appears to stay intact, what I was actually seeing was just that (sometimes) the _author_ stayed intact. But the _last modified timestamp_ for the other paragraphs in the outline does not - it is still updated to the current time even for unmodified paragraphs.

The reason I thought it was being retained previously was because in some cases (most, I suppose) if a page had multiple authors and I made a change via plugin, author info for the page was wiped out (along with "last modified" date and the page history that goes with it). So when I saw that sometimes I would make a change to a page via plugin and the authors' initials for other authors who had previously edited the page didn't disappear, I thought all such metadata was remaining intact. But after taking a closer look at the XML, that isn't the case - only the author is. I still don't know why the author is sometimes retained and sometimes not, but the last modified time never is.

In any case, I really hope MS fixes this eventually as the page history is a really amazing feature that is basically totally broken because of this (if I had to pick one bug for them to fix, this would probably be it...).
